### PR TITLE
[Feature] - Return the response object from queryAPI even when an error occurs

### PR DESCRIPTION
**Pull request by twentymurial33**
Implement the returned object to return the response object from `queryAPI` even when an error occurs 
**The goal is to enable more robust error-handling solutions by returning the error message.**

https://developer.mozilla.org/en-US/docs/Web/API/fetch

### CODE REVIEW

**Review by brentmclark — COMMENTED** [review]
Line `21` of this file still returns the result object.  Can you remove that and let the last line do the returning?

File: client/src/util/fetchFromAPI.js [hunk 1]
```diff
@@ -22,4 +22,5 @@ export const queryAPI = async (options, toast) => {
   } else {
     toast.error(`Food item not saved.  Error: ${result.error}`);
   }
+  return response;
```
Comment (brentmclark) [inline client/src/util/fetchFromAPI.js:25]: Ideally `queryAPI` would return the `result` object, not the `response` object.  This also means that you'll have to set `result.error` inside of the `else` branch of the condition.

File: client/src/util/fetchFromAPI.js [hunk 2]
```diff
@@ -2,13 +2,13 @@ export const queryAPI = async (options, toast) => {
...
+  const { url, onError, method } = options;
   if (url === undefined) {
     throw new Error("url is not provided");
   }
-  if (onSuccess === undefined) {
-    throw new Error("onSuccess is not provided");
-  }
+  // if (onSuccess === undefined) {
+  //   throw new Error("onSuccess is not provided");
+  // }
```
[the remark this answers is not in the document]
Comment (twentymurial33) [reply client/src/util/fetchFromAPI.js:11]: yes sir

File: client/src/util/fetchFromAPI.js (hunk 1 above)
Comment (twentymurial33) [reply client/src/util/fetchFromAPI.js]: thank you -updated

**Comment by twentymurial33** [comment]
> Line `21` of this file still returns the result object. Can you remove that and let the last line do the returning?

sorry about that ..took the notes during our call and yet missed some 👎

File: client/src/util/fetchFromAPI.js [hunk 3]
```diff
@@ -18,8 +16,9 @@ export const queryAPI = async (options, toast) => {
   };
   const response = await fetch(url, method);
   if (response.ok) {
-    return (result.data = await response.json());
+    result.data = await response.json();
   } else {
-    toast.error(`Food item not saved.  Error: ${result.error}`);
+    result.error(`Food item not saved.  Error: ${result.error}`);
```
Comment (brentmclark) [inline client/src/util/fetchFromAPI.js:21]: Did you mean to change this from `toast` to result`?  It's probably a good time to over this branch of the condition with a test so you can validate that it works as expected.  How do you feel about tackling that?

Comment (twentymurial33) [reply client/src/util/fetchFromAPI.js:21]: I'm good to give it a try 💯

Comment (brentmclark) [reply client/src/util/fetchFromAPI.js:21]: Let's focus our efforts here for a little bit.  Taking the convo to Slack though.

Comment (twentymurial33) [reply client/src/util/fetchFromAPI.js:21]: Screen Shot 2023-02-14 at 9 44 42 AM

Pushed new changes - I'm I heading in the right direction? Currently failing with error no options passed

Comment (brentmclark) [reply client/src/util/fetchFromAPI.js:21]: You've got a test in place, but it's not testing the thing I had in mind.  I'm clearly doing a terrible job expressing myself 😞 

The test should do the following

- Force `fetch` to receive a non-200 response (from the mock)
- verify that `queryAPI` returns an object with an `error` property
- verify that the `error` property is not null
- verify that the `error` property contains the response message from the server

That first bullet point can be extra tricky.  Have a look at the `jest-fetch-mock` documentation for mocking a failed fetch.  https://github.com/jefflau/jest-fetch-mock#mocking-a-failed-fetch

Comment (twentymurial33) [reply client/src/util/fetchFromAPI.js:21]: Screen Shot 2023-02-16 at 9 04 06 PM

Comment (twentymurial33) [reply client/src/util/fetchFromAPI.js:21]: Totally forgot to git push ...pushed for review - sorry that was just sitting somewhere 👎

File: client/src/util/fetchFromAPI.test.js [hunk 4]
```diff
@@ -32,15 +32,16 @@ it("mocks the fetch request", async () => {
...
-      expect(fetch).toEqual(result.error);
-    });
+    const onSuccess = jest.fn();
+    const onError = jest.fn();
+    return queryAPI("/")
+      .then(onSuccess)
+      .catch(onError)
+      .finally(() => {
+        expect(onSuccess).not.toHaveBeenCalled();
+        expect(onError).toHaveBeenCalled();
```
Comment (brentmclark) [inline client/src/util/fetchFromAPI.test.js:44]: Validate that onError was called with the message `error message`.

Comment (twentymurial33) [reply client/src/util/fetchFromAPI.test.js]: @brentmclark pushed and test has passed ...I'm I testing the correctly or rather what is intended to be tested?

Screen Shot 2023-03-14 at 11 24 43 AM
Screen Shot 2023-03-14 at 11 24 52 AM

**Review by brentmclark — APPROVED** [review]
This is looking much better!  My confidence in the implementation would be higher if the test also validated that `onError` _was_ called in addition to `onSuccess` _not_ being called, but you've understood the basic premise of testing an ajax call and that's the point of this exercise.  We can iterate from here.

**Pull request merged**